demos/https/iot_demo_https_sync.c

```diff
@@ -0,0 +1,499 @@
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "Failed to write the header \"Range: bytes=0-0\" into the request. With error code: %d", 
+            httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Send the request synchronously. This will also create an implicit connection.*/
+    httpsClientStatus = IotHttpsClient_SendSync( &connHandle, reqHandle, &respHandle );
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "There has been an error receiving the response. The error code is: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* If the server could not response with a 206 Partial Content response, then there was a problem retrieving the
+       file size and we will exit the demo. */
+    IotHttpsClient_ReadResponseStatus(respHandle, &respStatus );
+    if( respStatus != IOT_HTTPS_STATUS_PARTIAL_CONTENT)
+    {
+        IotLogError("Could not retrieve the file size. s3 responded with response status: %d", respStatus);
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Get the file size by parsing the "bytes 0-0/FILESIZE" Content-Range header value string. */
+    httpsClientStatus = IotHttpsClient_ReadHeader( respHandle, 
+        CONTENT_RANGE_HEADER_FIELD, 
+        contentRangeValStr, 
+        sizeof(contentRangeValStr) );
+    fileSizeStr = strstr( contentRangeValStr, "/" );
+    if(fileSizeStr == NULL)
+    {
+        IotLogError("Expected the header value \"bytes 0-0/FILESIZE\" to be retrieved, but \"/\" could not be found in \
+        in the header value.");
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+
+    }
+    /* The file size string is the next character after the "/" */
+    fileSizeStr += 1;
+    fileSize = (uint32_t)strtoul(fileSizeStr, NULL, 10);
+
+    /* The number of bytes we want to request each time is the size of the buffer or the file size if it is smaller than 
+       the buffer size, then the size of the file. */
+    numReqBytes = IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE;
+    if( fileSize < IOT_DEMO_HTTPS_RESP_BODY_BUFFER_SIZE)
+    {
+        numReqBytes = fileSize;
+    }
+
+    /* Here we iterate sending byte range requests until the full file has been downloaded. We keep track of the next 
+       byte to download with curByte. When this reaches the fileSize we stop downloading.
+       */
+    while( curByte < fileSize)
+    {
+        /* Re-initialize the request to reuse the request. If we do not reinitialize then data from the last response
+           associated with this request will linger. We reuse reqHandle beccause we are sending a new sequential 
+           synchronous request. IotHttpsClient_InitializeRequest will create a new request from the reqConfig and return
+           a reqHandle that is ready to use as a NEW request. */
+        httpsClientStatus = IotHttpsClient_InitializeRequest( &reqHandle, &reqConfig );
```

Comment:
It does not reset the buffer to 0s. It only resets internal pointers to the start of the header space.

Reply:
I updated the API docs:
```
 * This function will initialize the input request configuration by setting where to write the next headers to the start
 * of the configured header buffer in #IotHttpsRequestInfo_t.reqUserBuffer. It will also initialize the input associated 
 * response such that it can be received at the start of the configured header buffer in 
 * #IotHttpsRequestInfo_t.respUserBuffer.
```

Reply:
Regarding not setting buffers to 0, is there a risk of leaking information from 1st request or its response to 2nd request or its response?

Reply:
After reading data from the network, the length of the buffer specified to parse is the full size of the buffer. I could have read less than the full size from the network. If I did read less the bufffer size from the network, then data from the last response may still be present after the data read. 
I will zero out the buffer to keep the parser from moving past the end of the message and possibly reading a body when there was none.